### DISCUSSION TITLE
Virtual Interfaces

### DISCUSSION THREAD

**Discussion by LupusInferni315** [dotnet/csharplang#6937] [Language Ideas]
I recently ran into a situation where I needed an interface `IFoo`, that inherited from and implemented another interface `IBar` with each implementer of `IFoo` being able to have public access to `IBar`s implemented method (which doesn't work without some sketchy code)

How this currently can work

```csharp
// T in this example is not generic, in the sense of c#, its simply to express the concept without a specific use case.
public interface IFoo : IBar {
    T Foo();
    public T Bar() { /* Some implementation */ }
    T IBar.Bar() => Bar();
}

// and each implementer would have to contain the following line, in addition to implementing Foo().
public T Bar() => (this as IFoo).Bar();
````

With the proposed virtual interface, the same concept would look like this

```csharp
public virtual interface IFoo : IBar {
    T Foo(); // same as interface
    public virtual T Bar() { // this would be inherited by the implementing class and would not require additional implementation.
    // Access Modifiers for these methods would be restricted from using internal as context would be ambiguous.
        /* Some implementation */ 
    }
}
```

**Comment by siegfriedpammer** [comment]
Looks like this should be solved by using abstract classes.

**Reply by LupusInferni315** [reply]
With an important distinction, structs can only implement interfaces, which this virtual interface would fit, and while to at least some degree I understand why a struct cannot inherit 'stored' information (which a virtual interface would not contain), there is no good reason, why it must not inherit 'mechanics' information, so unless they allow structs to inherit structs the abstract idea won't really work here

**Comment by HaloFour** [comment]
This would be more of a runtime request.  The language doesn't control how the members do (or do not) appear to be inherited on the implementing types.

However, the current behavior of default implementations is based on the concept that interfaces are implemented, not inherited.  Implementing the contract doesn't require making those members accessible, as was already the case with explicit implementation since C# 1.0.  All the runtime cares about is that the members can be invoked through the contract.  Default implementations can be thought of in the way of explicit implementation, except that the member isn't defined on the implementing class at all.

All that said I get why it's a nuisance to not be able to invoke the members directly on the implementing class, and I think the inability to do so makes it a challenge to use interfaces like traits in this way.

**Reply by LupusInferni315** [reply]
I would agree with you, and the previous answer, but specifically with structs you cannot inherit from anything other than an interface (hence the suggestion of a virtual interface from which a class/struct could implement and optionally override the member but would otherwise inherit the existing one.) As for the runtime request portion of your answer, I wouldn't necessarily know what needs to be done compiler wise to actually achieve this and thus not entirely certain as to what type of request I am actually making.

**Reply by Joe4evr** [reply]
Perhaps you should rethink your use of structs, then. If you're not writing the implementation of an interface member in the struct itself, you'll be paying boxing/unboxing cost when invoking it.

**Reply by LupusInferni315** [reply]
The specific case that led to my proposal, I wouldn't be able to limit implementations to class, as it's part of a library where the user would define all types implementing the interface. The interface is used in a method that takes any type and if the type implements the interface to take it into account.
